Mathlib/RingTheory/PowerSeries/Substitution.lean

```diff
@@ -0,0 +1,381 @@
+noncomputable def subst (a : MvPowerSeries τ S) (f : PowerSeries R) :
+    MvPowerSeries τ S :=
+  MvPowerSeries.subst (fun _ ↦ a) f
+
+variable {a : MvPowerSeries τ S} {b : S⟦X⟧}
+
+/-- Substitution of power series into a power series -/
+noncomputable def substAlgHom (ha : HasSubst a) :
+    PowerSeries R →ₐ[R] MvPowerSeries τ S :=
+  MvPowerSeries.substAlgHom ha.const
+
+theorem coe_substAlgHom (ha : HasSubst a) :
+    ⇑(substAlgHom ha) = subst (R := R) a :=
+  MvPowerSeries.coe_substAlgHom ha.const
+
+/-- Rewrite `PowerSeries.substAlgHom` as `PowerSeries.aeval`.
+
+Its use is discouraged because it introduces a topology and might lead
+into awkward comparisons. -/
+theorem substAlgHom_eq_aeval
+    [UniformSpace R] [DiscreteUniformity R] [UniformSpace S] [DiscreteUniformity S]
+    (ha : HasSubst a) :
+    (substAlgHom ha : R⟦X⟧ →ₐ[R] MvPowerSeries τ S) = PowerSeries.aeval ha.hasEval := by
+  ext1 f
+  simpa only [substAlgHom, aeval, MvPowerSeries.coe_substAlgHom]
+    using congr_fun (MvPowerSeries.substAlgHom_eq_aeval (R := R) ha.const) f
+
+theorem subst_add (ha : HasSubst a) (f g : PowerSeries R) :
+    subst a (f + g) = subst a f + subst a g := by
+  rw [← coe_substAlgHom ha, map_add]
+
+theorem subst_pow (ha : HasSubst a) (f : PowerSeries R) (n : ℕ) :
+    subst a (f ^ n) = (subst a f ) ^ n := by
+  rw [← coe_substAlgHom ha, map_pow]
+
+theorem subst_mul (ha : HasSubst a) (f g : PowerSeries R) :
+    subst a (f * g) = subst a f * subst a g := by
+  rw [← coe_substAlgHom ha, map_mul]
+
+theorem subst_smul [Algebra A S] [IsScalarTower A R S]
+    (ha : HasSubst a) (r : A) (f : PowerSeries R) :
+    subst a (r • f) = r • (subst a f) := by
+  rw [← coe_substAlgHom ha, AlgHom.map_smul_of_tower]
+
+theorem coeff_subst_finite (ha : HasSubst a) (f : PowerSeries R) (e : τ →₀ ℕ) :
+    Set.Finite (fun (d : ℕ) ↦ (coeff R d f) • (MvPowerSeries.coeff S e (a ^ d))).support := by
+  convert (MvPowerSeries.coeff_subst_finite ha.const f e).image
+    ⇑(Finsupp.LinearEquiv.finsuppUnique ℕ ℕ Unit).toEquiv
+  rw [← Equiv.preimage_eq_iff_eq_image, ← Function.support_comp_eq_preimage]
+  apply congr_arg
+  rw [← Equiv.eq_comp_symm]
+  ext d
+  simp only [Finsupp.prod_pow, Finset.univ_unique, PUnit.default_eq_unit, Finset.prod_singleton,
+    LinearEquiv.coe_toEquiv_symm, EquivLike.coe_coe, Function.comp_apply,
+    Finsupp.LinearEquiv.finsuppUnique_symm_apply, Finsupp.single_eq_same]
+  congr
+
+theorem coeff_subst_finite' (hb : HasSubst b) (f : PowerSeries R) (e : ℕ) :
+    Set.Finite (fun (d : ℕ) ↦ (coeff R d f) • (PowerSeries.coeff S e (b ^ d))).support :=
+  coeff_subst_finite hb f  _
+
+theorem coeff_subst (ha : HasSubst a) (f : PowerSeries R) (e : τ →₀ ℕ) :
+    MvPowerSeries.coeff S e (subst a f) =
+      finsum (fun (d : ℕ) ↦
+        (coeff R d f) • (MvPowerSeries.coeff S e (a ^ d))) := by
+  erw [MvPowerSeries.coeff_subst ha.const f e]
+  rw [← finsum_comp_equiv (Finsupp.LinearEquiv.finsuppUnique ℕ ℕ Unit).toEquiv.symm]
```

Comment:
```suggestion
  rw [subst, MvPowerSeries.coeff_subst ha.const f e, ← finsum_comp_equiv
    (Finsupp.LinearEquiv.finsuppUnique ℕ ℕ Unit).toEquiv.symm]
```
In general let's try to avoid `erw` (we have `erw?` to debug it).